### PR TITLE
Generic Provider

### PR DESCRIPTION
**Pull request by rmeissner**
- Provider that can be used with common web3 javacript libraries
- Add transaction receipt type

- Based on https://github.com/Uniswap/uniswap-interface/compare/main...rimeissner:feature/safe_app

### Dependencies
- The new package depends on a change in an existing package (for the transaction receipt). Does this require to make it 2 PRs?

### CODE REVIEW

**Comment by rmeissner** [comment]
We enforce `;` at the end of the line? ... Who does that :P will adjust this later

**Comment by mikheevm** [comment]
> We enforce `;` at the end of the line? ... Who does that :P will adjust this later

Not us really, it's prettier plugin

File: packages/safe-apps-provider/src/provider.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+  send(request: any, callback: (error: any, response?: any) => void): void {
+    if (!request) callback('Undefined request');
+    this.request(request)
+      .then((result) => callback(null, { jsonrpc: '2.0', id: request.id, result }))
+      .catch((error) => callback(error, null));
+  }
+
+  async request(request: { method: string; params: any[] }): Promise<any> {
+    console.error(request.method, request.params);
```
Comment (rmeissner) [inline packages/safe-apps-provider/src/provider.ts:43]: This should be removed

File: packages/safe-apps-provider/src/provider.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    switch (request.method) {
+      case 'eth_accounts':
+        return [this.safe.safeAddress];
+
+      case 'net_version':
+      case 'eth_chainId':
+        return `0x${this.chainId.toString(16)}`;
+
+      case 'eth_sendTransaction':
+        console.error({ params });
```
Comment (rmeissner) [inline packages/safe-apps-provider/src/provider.ts:54]: Same here

File: packages/safe-apps-sdk/package.json [hunk 3]
```diff
@@ -15,7 +15,7 @@
   ],
   "scripts": {
     "test": "jest",
-    "format-dist": "sed -i '' 's/\"files\":/\"_files\":/' dist/package.json",
+    "format-dist": "sed -i'' 's/\"files\":/\"_files\":/' dist/package.json",
```
Comment (rmeissner) [inline packages/safe-apps-sdk/package.json:18]: does this work for you on mac @mikheevm ?

Comment (mikheevm) [reply packages/safe-apps-sdk/package.json:18]: https://stackoverflow.com/questions/4247068/sed-command-with-i-option-failing-on-mac-but-works-on-linux

Comment (mikheevm) [reply packages/safe-apps-sdk/package.json:18]: yes it works

Comment (mikheevm) [reply packages/safe-apps-sdk/package.json:18]: on linux you have to remove `''`

Comment (mikheevm) [reply packages/safe-apps-sdk/package.json:18]: https://github.com/gnosis/safe-apps-sdk/tree/master/packages/safe-apps-sdk#build

Comment (rmeissner) [reply packages/safe-apps-sdk/package.json:18]: also worked by just removing the space :D

File: packages/safe-apps-provider/README.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+```bash
+yarn add @gnosis.pm/safe-apps-provider
+
+npm i @gnosis.pm/safe-apps-provider
+```
+
+### SafeAppsProvider
+
+- Usage example with [safe-apps-react-sdk](https://github.com/gnosis/safe-apps-sdk/tree/master/packages/safe-apps-react-sdk) and [Ethers.js](https://docs.ethers.io/v5/api/providers/other/#Web3Provider)
```
Comment (mikheevm) [inline packages/safe-apps-provider/README.md:19]: Took me some time to realize that the difference between those snippers are about web3/ethers, maybe add a heading to both of them like `Usage example with ethers.js`

File: packages/safe-apps-react-sdk/package.json [hunk 5]
```diff
@@ -11,7 +11,7 @@
   ],
```
Comment (mikheevm) [inline packages/safe-apps-react-sdk/package.json:11]: I think for react sdk you have to change the version too

**Review by mikheevm — APPROVED** [review]

**Pull request merged**